### PR TITLE
nilrt-grub-runmode: Enable global toggling of mitigations for CPU vulnerabilities

### PR DESCRIPTION
**Pull request by ckoellin**
Sumo enabled toggling various spectre/meltdown mitigations related to CPU vulnerabilities. As time has moved on and we have upgraded to newer kernels there have been a number of additional [mitigation options](https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/kernel-parameters.txt#L3192) added. It is unrealistic to have to release a new GRUB fragment for toggling newer mitigations every time one is added. Instead we should use the global mitigations kernel boot parameter to do the toggling. This kernel boot parameter is updated to cover newer mitigations as they are added in newer kernels. Thus not requiring us to re-release.

### Testing

- I was able to build and deploy nilrt-grub-runmode to a NILRT target and was able to verify the installed file.
nilrt-grub-runmode

- I was also able to verify the mitigations were correctly enabled or disabled by renaming cpu-mitigations.cfg on hardknott
```
// cpu-mitigations.cfg
# grep . /sys/devices/system/cpu/vulnerabilities/*
/sys/devices/system/cpu/vulnerabilities/itlb_multihit:Not affected
/sys/devices/system/cpu/vulnerabilities/l1tf:Not affected
/sys/devices/system/cpu/vulnerabilities/mds:Not affected
/sys/devices/system/cpu/vulnerabilities/meltdown:Vulnerable
/sys/devices/system/cpu/vulnerabilities/mmio_stale_data:Not affected
/sys/devices/system/cpu/vulnerabilities/spec_store_bypass:Vulnerable
/sys/devices/system/cpu/vulnerabilities/spectre_v1:Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
/sys/devices/system/cpu/vulnerabilities/spectre_v2:Vulnerable, STIBP: disabled
/sys/devices/system/cpu/vulnerabilities/srbds:Not affected
/sys/devices/system/cpu/vulnerabilities/tsx_async_abort:Not affected
```
```
// cpu-mitigations.cfg.disable
# grep . /sys/devices/system/cpu/vulnerabilities/*
/sys/devices/system/cpu/vulnerabilities/itlb_multihit:Not affected
/sys/devices/system/cpu/vulnerabilities/l1tf:Not affected
/sys/devices/system/cpu/vulnerabilities/mds:Not affected
/sys/devices/system/cpu/vulnerabilities/meltdown:Mitigation: PTI
/sys/devices/system/cpu/vulnerabilities/mmio_stale_data:Not affected
/sys/devices/system/cpu/vulnerabilities/spec_store_bypass:Vulnerable
/sys/devices/system/cpu/vulnerabilities/spectre_v1:Mitigation: usercopy/swapgs barriers and __user pointer sanitization
/sys/devices/system/cpu/vulnerabilities/spectre_v2:Mitigation: Retpolines, STIBP: disabled, RSB filling
/sys/devices/system/cpu/vulnerabilities/srbds:Not affected
/sys/devices/system/cpu/vulnerabilities/tsx_async_abort:Not affected
```

### CODE REVIEW

**Review by shruthi-ravi — APPROVED** [review]
Sorry if this has already been discussed and I'm asking again. What can a user do now to enable a few and disable the remaining mitigations?

File: recipes-bsp/grub/grub/grub-runmode-bootimage.cfg [hunk 1]
```diff
@@ -5,16 +5,10 @@ set kernel_path='/runmode/bzImage'
...
-# into /runmode/bootimage.d. We also cannot load them via wildcard
-# (e.g. cve-*.cfg) because the provisioned grub might not have the
-# regexp module that adds support for file globbing.
-for mitigation_cfg in cve-2017-5715 cve-2017-5754 cve-2018-3620_3646 cve-2018-3639; do
-	if [ -f "/runmode/${mitigation_cfg}.cfg" ]; then
-		source "/runmode/${mitigation_cfg}.cfg"
-	fi
-done
+# into /runmode/bootimage.d.
+source "/runmode/cpu-mitigations.cfg"
```
Comment (bstreiff) [inline recipes-bsp/grub/grub/grub-runmode-bootimage.cfg]: if we're still doing the "rename to `.cfg.disable`" thing then this `source` should be guarded with an `if [ -f ... ]`

**Comment by bstreiff** [comment]
> What can a user do now to enable a few and disable the remaining mitigations?

They can edit the boot arguments in bootimage.cfg.

Adding support for the `nirtcfg` backend to manipulate the .cfg files that control individual mitigations mostly makes sense from the standpoint of "then it can be plumbed up to be a checkbox in MAX" (I'm not sure if this was explored); this was perhaps a more tenable idea back when there were just four of them, but upstream's up to eleven mitigation options now. From both a UX and a maintainability standpoint it's far more reasonable to expose an easy-mode for the global `mitigations` option, and let individual control be an advanced use case.

**Comment by ckoellin** [comment]
> They can edit the boot arguments in bootimage.cfg.

To my knowledge our [documentation ](https://www.ni.com/content/ni/locales/en-us/support/documentation/supplemental/19/enabling-security-mitigations-for-meltdown-and-spectre-on-ni-lin.html) only points them to use nirtcfg. We should probably update this documentation with an advanced section of some kind to say if they really need the control what mitigations are enabled they can edit cpu-mitigations.cfg or append to otherbootargs via bootimage.cfg or another grub fragment.

File: recipes-bsp/grub/grub/grub-runmode-bootimage.cfg (hunk 1 above)
Comment (ckoellin) [reply recipes-bsp/grub/grub/grub-runmode-bootimage.cfg]: Updated.

**Review by amstewart — APPROVED** [review]

**Review by gratian — APPROVED** [review]

**Pull request merged**